File: ios/CashuWallet/Core/NostrMintBackupService.swift
import Foundation
import Cdk

/// Publishes the wallet's mint list as an encrypted NUT-27 backup on Nostr and
/// finds existing backups during restore. All protocol work (key derivation,
/// encryption, relay publishing) happens inside cdk via `WalletRepository`;
/// this service only adds settings gating, relay hygiene, and UI-facing state.
@MainActor
final class NostrMintBackupService: ObservableObject {
    static let shared = NostrMintBackupService()

    @Published private(set) var isBackingUp = false
    @Published private(set) var isSearching = false
    @Published private(set) var lastBackupDate: Date?

    /// Injected by `WalletManager` alongside its own repository lifecycle.
    var walletRepository: WalletRepository?

    private init() {
        lastBackupDate = UserDefaults.standard.object(forKey: StorageKeys.nostrMintBackupLastBackupDate) as? Date
    }

    /// Fire-and-forget trigger after mint-list changes — the Nostr twin of
    /// `performICloudBackup()`. Failures only log; the mint operation that
    /// triggered the backup must not surface a relay error.
    func backupCurrentMintsIfEnabled() async {
        guard SettingsManager.shared.nostrMintBackupEnabled else { return }
        do {
            try await backupMints()
        } catch NostrMintBackupError.nothingToBackUp {
            // Empty wallet — nothing worth publishing, not a failure.
        } catch {
            AppLogger.wallet.error("Nostr mint backup failed: \(error)")
        }
    }

    func backupMints() async throws {
        guard SettingsManager.shared.useWebsockets else {
            throw NostrMintBackupError.webSocketsDisabled
        }
        guard let walletRepository else {
            throw NostrMintBackupError.notInitialized
        }
        let relays = normalizedRelays(SettingsManager.shared.nostrRelays)
        guard !relays.isEmpty else {
            throw NostrMintBackupError.noRelays
        }

        // NUT-27 backups are addressable events: publishing replaces the
        // previous backup for this seed on the relay. Never push an empty
        // list — a freshly initialized wallet (e.g. mid-restore) would
        // otherwise wipe the backup it is about to read.
        guard await !walletRepository.getWallets().isEmpty else {
            throw NostrMintBackupError.nothingToBackUp
        }

        isBackingUp = true
        defer { isBackingUp = false }

        _ = try await walletRepository.backupMints(
            relays: relays,
            options: BackupOptions(client: "cashu.me")
        )

        let date = Date()
        lastBackupDate = date
        UserDefaults.standard.set(date, forKey: StorageKeys.nostrMintBackupLastBackupDate)
    }

    /// Fetch the newest mint-list backup for the currently opened wallet seed.
    /// Returns the backed-up mint URLs (empty when the relays have no backup).
    func fetchBackedUpMintURLs() async throws -> [String] {
        guard SettingsManager.shared.useWebsockets else {
            throw NostrMintBackupError.webSocketsDisabled
        }
        guard let walletRepository else {
            throw NostrMintBackupError.notInitialized
        }
        let relays = normalizedRelays(SettingsManager.shared.nostrRelays)
        guard !relays.isEmpty else {
            throw NostrMintBackupError.noRelays
        }

        isSearching = true
        defer { isSearching = false }

        let backup = try await walletRepository.fetchMintBackup(
            relays: relays,
            options: RestoreOptions(timeoutSecs: 4)
        )
        return backup.mints.map(\.url)
    }

    func resetForWalletBoundary() {
        lastBackupDate = nil
        UserDefaults.standard.removeObject(forKey: StorageKeys.nostrMintBackupLastBackupDate)
    }

    private func normalizedRelays(_ relays: [String]) -> [String] {
        var seen = Set<String>()
        return relays.compactMap { relay in
            let trimmed = relay.trimmingCharacters(in: .whitespacesAndNewlines)
            let lower = trimmed.lowercased()
            guard lower.hasPrefix("wss://") || lower.hasPrefix("ws://") else { return nil }
            guard seen.insert(trimmed).inserted else { return nil }
            return trimmed
        }
    }
}

enum NostrMintBackupError: LocalizedError {
    case notInitialized
    case noRelays
    case webSocketsDisabled
    case nothingToBackUp

    var errorDescription: String? {
        switch self {
        case .notInitialized:
            return "Wallet is not initialized."
        case .noRelays:
            return "No Nostr relays are configured."
        case .webSocketsDisabled:
            return "Websocket connections are disabled."
        case .nothingToBackUp:
            return "There are no mints to back up yet."
        }
    }
}

File: android/app/src/main/java/com/cashu/me/Core/NostrMintBackupService.kt
package com.cashu.me.Core

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import com.cashu.me.Core.CDK.CdkWalletGateway

data class NostrMintBackupState(
    val isBackingUp: Boolean = false,
    val isSearching: Boolean = false,
    val lastBackupDateEpochMillis: Long? = null,
)

sealed class NostrMintBackupException(message: String) : IllegalStateException(message) {
    class WebSocketsDisabled : NostrMintBackupException("Websocket connections are disabled.")
    class NoRelays : NostrMintBackupException("No Nostr relays are configured.")
    class NothingToBackUp : NostrMintBackupException("There are no mints to back up yet.")
}

/**
 * Publishes the wallet's mint list as an encrypted NUT-27 backup on Nostr and
 * finds existing backups during restore. All protocol work (key derivation,
 * encryption, relay publishing) happens inside cdk via the wallet repository;
 * this service only adds settings gating, relay hygiene, and UI-facing state.
 * Mirrors iOS `NostrMintBackupService`.
 */
class NostrMintBackupService(
    private val settingsManager: SettingsManager,
    private val settingsStore: SettingsStore,
    private val gateway: CdkWalletGateway,
) {
    private val mutableState = MutableStateFlow(
        NostrMintBackupState(lastBackupDateEpochMillis = settingsStore.nostrMintBackupLastBackupDate),
    )
    val state: StateFlow<NostrMintBackupState> = mutableState.asStateFlow()

    /**
     * Fire-and-forget trigger after mint-list changes — the Nostr twin of the
     * iCloud backup on iOS. Failures only log; the mint operation that
     * triggered the backup must not surface a relay error.
     */
    suspend fun backupCurrentMintsIfEnabled() {
        if (!settingsManager.state.value.nostrMintBackupEnabled) return
        try {
            backupMints()
        } catch (_: NostrMintBackupException.NothingToBackUp) {
            // Empty wallet — nothing worth publishing, not a failure.
        } catch (t: Throwable) {
            AppLogger.wallet.error("Nostr mint backup failed", t)
        }
    }

    suspend fun backupMints() {
        val relays = requireBackupPreconditions()

        // NUT-27 backups are addressable events: publishing replaces the
        // previous backup for this seed on the relay. Never push an empty
        // list — a freshly initialized wallet (e.g. mid-restore) would
        // otherwise wipe the backup it is about to read.
        if (!gateway.hasWallets()) throw NostrMintBackupException.NothingToBackUp()

        mutableState.value = mutableState.value.copy(isBackingUp = true)
        try {
            gateway.backupMints(relays, client = BACKUP_CLIENT)
            val now = System.currentTimeMillis()
            settingsStore.nostrMintBackupLastBackupDate = now
            mutableState.value = mutableState.value.copy(lastBackupDateEpochMillis = now)
        } finally {
            mutableState.value = mutableState.value.copy(isBackingUp = false)
        }
    }

    /**
     * Fetch the newest mint-list backup for the currently opened wallet seed.
     * Returns the backed-up mint URLs (empty when the relays have no backup).
     */
    suspend fun fetchBackedUpMintUrls(): List<String> {
        val relays = requireBackupPreconditions()
        mutableState.value = mutableState.value.copy(isSearching = true)
        return try {
            gateway.fetchMintBackup(relays, timeoutSecs = RESTORE_TIMEOUT_SECS)
        } finally {
            mutableState.value = mutableState.value.copy(isSearching = false)
        }
    }

    fun resetForWalletBoundary() {
        settingsStore.nostrMintBackupLastBackupDate = null
        mutableState.value = NostrMintBackupState()
    }

    fun reloadStoredState() {
        mutableState.value = NostrMintBackupState(
            lastBackupDateEpochMillis = settingsStore.nostrMintBackupLastBackupDate,
        )
    }

    private fun requireBackupPreconditions(): List<String> {
        val settings = settingsManager.state.value
        if (!settings.useWebsockets) throw NostrMintBackupException.WebSocketsDisabled()
        val relays = normalizedNostrBackupRelays(settings.nostrRelays)
        if (relays.isEmpty()) throw NostrMintBackupException.NoRelays()
        return relays
    }

    companion object {
        private const val BACKUP_CLIENT = "cashu.me"
        private val RESTORE_TIMEOUT_SECS = 4uL

        /** Trim, keep only ws(s):// relays, dedupe preserving order (iOS normalizedRelays). */
        fun normalizedNostrBackupRelays(relays: List<String>): List<String> = relays
            .map(String::trim)
            .filter { it.startsWith("wss://", ignoreCase = true) || it.startsWith("ws://", ignoreCase = true) }
            .distinct()
    }
}
